### DISCUSSION TITLE
question about the Transformation API

### DISCUSSION THREAD

**Discussion by YeMengHua417** [iTwin/itwinjs-core#3924] [Q&A] [answered]
My understanding for the workflow is : we could “copy ” part of  the original iModel component,like model,spatial category,or filter some part of iModel to the target iModel, please correct me if I am wrong~ 
And My steps are 
1 prepare two projects with iModel,  the source iModel is from a dgn file ,and the second iModel is empty
2 call the https://developer.bentley.com/apis/transformations/operations/combineimodels/#combineimodels ,the response returns 201

{
    "configuration":{
        "transformParameters":{
            "iModels":[
                {
                    "_links":{
                        "project":{
                            "href":"https://api.bentley.com/projects/44c04181-9af0-4b1b-bace-8008054db97f"
                        },
                        "iModel":{
                            "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb"
                        }
                    }
                }
            ]
        },
        "comment":"Example comment",
        "createdDateTime":"2022-07-13T01:26:47.8751262Z",
        "modifiedDateTime":"2022-07-13T01:26:47.8733333Z",
        "_links":{
            "sourceIModel":{
                "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb"
            },
            "targetIModel":{
                "href":"https://api.bentley.com/imodels/34947f3d-54b7-47f1-a074-db4eb187e426"
            },
            "sourceProject":{
                "href":"https://api.bentley.com/projects/44c04181-9af0-4b1b-bace-8008054db97f"
            },
            "targetProject":{
                "href":"https://api.bentley.com/projects/e9ffee88-6499-40f2-84e0-57d64c64ef48"
            }
        },
        "id":"3cbbee61-e0a4-469e-5ef4-08da5b568de5",
        "transformName":"Example name"
    }
}

3 call the https://developer.bentley.com/apis/transformations/operations/create-transformation/#create-transformation, by only modify the configurationId to "3cbbee61-e0a4-469e-5ef4-08da5b568de5"
But the response is 422

api-supported-versions: 2.0
content-type: application/json
date: Wed, 13 Jul 2022 01:40:38 GMT
request-context: appId=cid-v1:caffab81-b80a-40d1-a9e3-3d1a07228483
x-correlation-id: d5d100fe-75ec-47a9-bbf5-4a0396ffd46d
x-itwinplatform-media-type: bentley.itwin-platform.v1
x-itwinplatform-region: East US

{
    "error": {
        "code": "InvalidTransformationsRequest",
        "message": "Cannot create Transformation."
    }
}

 
Could you kindly point out which step is wrong ? or did I missing anything? thanks

**Comment by mindaugasdirg** [comment]
CombineIModels transformation doesn't allow to achieve that. It allows to combine multiple *full* iModels into a single target iModel. To copy *part* of source iModel into a target iModel you should use one of these transformations: [FilterByViewDefinition](https://developer.bentley.com/apis/transformations/operations/filterbyviewdefinition/), [FilterIModel](https://developer.bentley.com/apis/transformations/operations/filterimodel/) or [FilterSubCategories](https://developer.bentley.com/apis/transformations/operations/filtersubcategories/).

The steps are correct just in step 3 you are setting `sourceChangeSetId` property to `'0'` in https://developer.bentley.com/apis/transformations/operations/create-transformation/#create-transformation endpoint, but the source iModel that you are using doesn't seem to have any change sets with that ID.

 https://developer.bentley.com/apis/transformations/operations/create-transformation/#create-transformation accepts optional `sourceChangeSetId` property to allow transforming an older version of iModel in cases when not the latest changeset should be transformed. One such case could be a case when the newest named version should be transformed. Since the property is optional you can omit it and the latest changeset will be used automatically during the transformation.

Also to configure transformation you are not required to have two separate projects. All iModels can be in the same project.

**Comment by YeMengHua417** [comment]
Hi, thanks for the quick answer, I called the https://developer.bentley.com/apis/imodels/operations/get-imodel-changesets/
to get the latest changeset id by source iModelId afb0931a-1c36-4c67-b188-dec733751cdb
the Response is like the following ,and I retried the create transformation API by replacing the sourceChangeSetId byt the following changesetId ,but the response is still 422?
the request body for create-transformation
{
    "configurationId": "3cbbee61-e0a4-469e-5ef4-08da5b568de5",
    "sourceChangeSetId": "ffedfa1e4f0b5f5c0a6e987e4ad824972c1895c1"  ---from the following response
}
the Response from get-imodel-changesets:
{
    "changesets":[
        {
            "id":"4681a740b4d10e171d885a83bf3d507edada91cf",
            "displayName":"1",
            "description":"MicroStation Connector - Domain schema upgrade",
            "index":1,
            "parentId":"",
            "briefcaseId":2,
            "fileSize":6384,
            "state":"fileUploaded",
            "containingChanges":1,
            "creatorId":"f1efbd9a-2a93-4c30-8f82-b0cb06f05211",
            "pushDateTime":"2022-07-06T02:32:45.17Z",
            "_links":{
                "self":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/changesets/4681a740b4d10e171d885a83bf3d507edada91cf"
                },
                "creator":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/users/f1efbd9a-2a93-4c30-8f82-b0cb06f05211"
                }
            }
        },
        {
            "id":"db3c0e50fad288ad5af7ccfe53725de4c9876153",
            "displayName":"2",
            "description":"MicroStation Connector - iModelHub - Domain schema upgrade",
            "index":2,
            "parentId":"4681a740b4d10e171d885a83bf3d507edada91cf",
            "briefcaseId":2,
            "fileSize":3791,
            "state":"fileUploaded",
            "containingChanges":1,
            "creatorId":"f1efbd9a-2a93-4c30-8f82-b0cb06f05211",
            "pushDateTime":"2022-07-06T02:32:59.587Z",
            "_links":{
                "self":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/changesets/db3c0e50fad288ad5af7ccfe53725de4c9876153"
                },
                "creator":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/users/f1efbd9a-2a93-4c30-8f82-b0cb06f05211"
                }
            }
        },
        {
            "id":"d29688c77501d0c671f192522d4859a152c7d468",
            "displayName":"3",
            "description":"MicroStation Connector - iModelHub - BootstrapExternalSources",
            "index":3,
            "parentId":"db3c0e50fad288ad5af7ccfe53725de4c9876153",
            "briefcaseId":2,
            "fileSize":185,
            "state":"fileUploaded",
            "containingChanges":0,
            "creatorId":"f1efbd9a-2a93-4c30-8f82-b0cb06f05211",
            "pushDateTime":"2022-07-06T02:33:06.073Z",
            "_links":{
                "self":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/changesets/d29688c77501d0c671f192522d4859a152c7d468"
                },
                "creator":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/users/f1efbd9a-2a93-4c30-8f82-b0cb06f05211"
                }
            }
        },
        {
            "id":"198c5f8b5f24ed789e7f4cfa1a37503dc2192db7",
            "displayName":"4",
            "description":"MicroStation Connector - iModelHub - Initialization changes",
            "index":4,
            "parentId":"d29688c77501d0c671f192522d4859a152c7d468",
            "briefcaseId":2,
            "fileSize":691,
            "state":"fileUploaded",
            "containingChanges":0,
            "creatorId":"f1efbd9a-2a93-4c30-8f82-b0cb06f05211",
            "pushDateTime":"2022-07-06T02:33:14.04Z",
            "_links":{
                "self":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/changesets/198c5f8b5f24ed789e7f4cfa1a37503dc2192db7"
                },
                "creator":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/users/f1efbd9a-2a93-4c30-8f82-b0cb06f05211"
                }
            }
        },
        {
            "id":"f10897d23557032117fb10e60ec8ad1e3f2c7199",
            "displayName":"5",
            "description":"MicroStation Connector - iModelHub - SynchronizationComplete",
            "index":5,
            "parentId":"198c5f8b5f24ed789e7f4cfa1a37503dc2192db7",
            "briefcaseId":2,
            "fileSize":206,
            "state":"fileUploaded",
            "containingChanges":0,
            "creatorId":"f1efbd9a-2a93-4c30-8f82-b0cb06f05211",
            "pushDateTime":"2022-07-06T02:33:19.197Z",
            "_links":{
                "self":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/changesets/f10897d23557032117fb10e60ec8ad1e3f2c7199"
                },
                "creator":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/users/f1efbd9a-2a93-4c30-8f82-b0cb06f05211"
                }
            }
        },
        {
            "id":"9c06caf4bee7cca448a9f1323b8e64ddda6063ef",
            "displayName":"6",
            "description":"MicroStation Connector - iModelHub - Definitions",
            "index":6,
            "parentId":"f10897d23557032117fb10e60ec8ad1e3f2c7199",
            "briefcaseId":2,
            "fileSize":1178,
            "state":"fileUploaded",
            "containingChanges":2,
            "creatorId":"f1efbd9a-2a93-4c30-8f82-b0cb06f05211",
            "pushDateTime":"2022-07-06T02:33:25.087Z",
            "_links":{
                "self":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/changesets/9c06caf4bee7cca448a9f1323b8e64ddda6063ef"
                },
                "creator":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/users/f1efbd9a-2a93-4c30-8f82-b0cb06f05211"
                }
            }
        },
        {
            "id":"112f36ee987a6c8af26da8fcf90095b412d22f06",
            "displayName":"7",
            "description":"MicroStation Connector - iModelHub - Definitions",
            "index":7,
            "parentId":"9c06caf4bee7cca448a9f1323b8e64ddda6063ef",
            "briefcaseId":2,
            "fileSize":12720,
            "state":"fileUploaded",
            "containingChanges":2,
            "creatorId":"f1efbd9a-2a93-4c30-8f82-b0cb06f05211",
            "pushDateTime":"2022-07-06T02:33:30.067Z",
            "_links":{
                "self":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/changesets/112f36ee987a6c8af26da8fcf90095b412d22f06"
                },
                "creator":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/users/f1efbd9a-2a93-4c30-8f82-b0cb06f05211"
                }
            }
        },
        {
            "id":"20f3b0f53080577a15b59950eed1eed33ef1560f",
            "displayName":"8",
            "description":"MicroStation Connector - iModelHub - Views and Models",
            "index":8,
            "parentId":"112f36ee987a6c8af26da8fcf90095b412d22f06",
            "briefcaseId":2,
            "fileSize":1410,
            "state":"fileUploaded",
            "containingChanges":16,
            "creatorId":"f1efbd9a-2a93-4c30-8f82-b0cb06f05211",
            "pushDateTime":"2022-07-06T02:33:32.317Z",
            "_links":{
                "self":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/changesets/20f3b0f53080577a15b59950eed1eed33ef1560f"
                },
                "creator":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/users/f1efbd9a-2a93-4c30-8f82-b0cb06f05211"
                }
            }
        },
        {
            "id":"fafa39f0b75a0bc85fc0cf03da2e137150f48e40",
            "displayName":"9",
            "description":"MicroStation Connector - iModelHub - Spatial data",
            "index":9,
            "parentId":"20f3b0f53080577a15b59950eed1eed33ef1560f",
            "briefcaseId":2,
            "fileSize":2690889,
            "state":"fileUploaded",
            "containingChanges":4,
            "creatorId":"f1efbd9a-2a93-4c30-8f82-b0cb06f05211",
            "pushDateTime":"2022-07-06T02:33:57.733Z",
            "_links":{
                "self":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/changesets/fafa39f0b75a0bc85fc0cf03da2e137150f48e40"
                },
                "creator":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/users/f1efbd9a-2a93-4c30-8f82-b0cb06f05211"
                }
            }
        },
        {
            "id":"3a27ad89a0f7560232043ac5221bdb6720ade4e8",
            "displayName":"10",
            "description":"MicroStation Connector - iModelHub - Global properties",
            "index":10,
            "parentId":"fafa39f0b75a0bc85fc0cf03da2e137150f48e40",
            "briefcaseId":2,
            "fileSize":620,
            "state":"fileUploaded",
            "containingChanges":32,
            "creatorId":"f1efbd9a-2a93-4c30-8f82-b0cb06f05211",
            "pushDateTime":"2022-07-06T02:34:02.597Z",
            "_links":{
                "self":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/changesets/3a27ad89a0f7560232043ac5221bdb6720ade4e8"
                },
                "creator":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/users/f1efbd9a-2a93-4c30-8f82-b0cb06f05211"
                }
            }
        },
        {
            "id":"ffedfa1e4f0b5f5c0a6e987e4ad824972c1895c1",
            "displayName":"11",
            "description":"MicroStation Connector - iModelHub - Finalization changes",
            "index":11,
            "parentId":"3a27ad89a0f7560232043ac5221bdb6720ade4e8",
            "briefcaseId":2,
            "fileSize":1558,
            "state":"fileUploaded",
            "containingChanges":32,
            "creatorId":"f1efbd9a-2a93-4c30-8f82-b0cb06f05211",
            "pushDateTime":"2022-07-06T02:34:13.557Z",
            "_links":{
                "self":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/changesets/ffedfa1e4f0b5f5c0a6e987e4ad824972c1895c1"
                },
                "creator":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/users/f1efbd9a-2a93-4c30-8f82-b0cb06f05211"
                }
            }
        },
        {
            "id":"618796ba9a93e98c4b1467c8c5ecd5f39a6392ea",
            "displayName":"12",
            "description":"MicroStation Connector - iModelHub - Finalization changes",
            "index":12,
            "parentId":"ffedfa1e4f0b5f5c0a6e987e4ad824972c1895c1",
            "briefcaseId":2,
            "fileSize":112,
            "state":"fileUploaded",
            "containingChanges":32,
            "creatorId":"f1efbd9a-2a93-4c30-8f82-b0cb06f05211",
            "pushDateTime":"2022-07-12T06:34:19.19Z",
            "_links":{
                "self":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/changesets/618796ba9a93e98c4b1467c8c5ecd5f39a6392ea"
                },
                "creator":{
                    "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/users/f1efbd9a-2a93-4c30-8f82-b0cb06f05211"
                }
            }
        }
    ],
    "_links":{
        "self":{
            "href":"https://api.bentley.com/imodels/afb0931a-1c36-4c67-b188-dec733751cdb/changesets?$skip=0&$top=100"
        },
        "prev":null,
        "next":null
    }
}

**Comment by mindaugasdirg** [comment, marked as answer]
Sorry for not responding sooner. I've looked into the request data you provided and I see that you have misconfigured the transformation by providing the same iModel twice - once as the main source iModel and the second time in the `transformParameters` property. You need to provide a different source iModel in the `transformParameters` which would be merged with the iModel specified in the `sourceIModelId` property. Or you should use another transformation's configuration since CombineIModels only aim is to merge multiple iModels into one iModel.

**Reply by YeMengHua417** [reply]
Many thanks for reply, got 200 return code finally. If I got a circle in source iModel and a triangle in transformParameters, does that mean I could get both figure in the target iModel? BTW,is there any limitations for the source iModel or iModel in the transformParameters? e.g ,like size or something else？

**Reply by mindaugasdirg** [reply]
Yes, CombineIModels transformation is exactly for this purpose. It will export everything from all source iModels (`sourceIModelId` and `transformParameter` list) into target iModel including geometry and all additional information present in source iModels. So the target iModel will have circle and triangle elements.

For your second question, Transformations API doesn't have any strict limitations on source iModels, however processing time scales with iModel's size so larger iModels can take a long time to fully process them. Processing might be stopped if it takes too long, but it's a rare occurrence.
